### PR TITLE
SE-448: Injection Alerts App Form

### PR DESCRIPTION
**Pull request by bkemper**
### What Changed
 - Allow creation of a "Injection Count" alert when "allow_injection_alerts" feature flag is enabled

### How To Test
 - If feature flag is enabled, allow user to create a new "Injection Count" alert
 - If feature flag is disabled, allow account to see and edit existing "Injection Count" alerts, but do not allow the creation of a new alert.

### CODE REVIEW

File: src/pages/alerts/components/AlertForm.container.js [hunk 1]
```diff
@@ -3,30 +3,33 @@ import { reduxForm, formValueSelector, getFormSyncErrors, getFormMeta } from 're
...
 import { selectAlertFormValues } from 'src/selectors/alerts';
 
 export default function withAlertForm(WrappedComponent) {
-
   const mapStateToProps = (state, props) => {
     const selector = formValueSelector(FORM_NAME);
     const { isDuplicate, isNewAlert } = props;
 
     return {
+      allowInjectionAlerts: selectCondition(isAccountUiOptionSet('allow_injection_alerts'))(state),
```
Comment (bkemper) [inline src/pages/alerts/components/AlertForm.container.js]: added.

File: src/pages/alerts/components/AlertForm.js [hunk 2]
```diff
@@ -6,25 +6,27 @@ import { TextFieldWrapper, SelectWrapper } from 'src/components';
 import FilterFields from './fields/FilterFields';
 import EvaluatorFields from './fields/EvaluatorFields';
 import SubaccountField from './fields/SubaccountsField';
-import getOptions from '../helpers/getOptions';
```
Comment (bkemper) [inline src/pages/alerts/components/AlertForm.js]: removed.

**Comment by bkemper** [comment]
@kchu1337 This is ready for the final pass.  The only thing left is to add some tests for `<SliderFieldWrapper />`.

File: src/pages/alerts/components/AlertForm.js [hunk 3]
```diff
@@ -65,63 +77,81 @@ export class AlertForm extends Component {
...
+      isDuplicate,
     } = this.props;
 
-    const submitText = submitting ? 'Submitting...' : (isNewAlert ? 'Create Alert' : 'Update Alert');
+    const submitText = submitting ? 'Submitting...' : isNewAlert ? 'Create Alert' : 'Update Alert';
     const isSubmitDisabled = (pristine && !isDuplicate) || submitting; //Allows user to create the same alert if if's a duplicate
     const formSpec = getFormSpec(metric);
     const channelsError = this.isNotificationChannelsEmpty(formMeta, formErrors);
+
+    const visibleMetricOptions = metricOptions.filter(option => {
```
Comment (bkemper) [inline src/pages/alerts/components/AlertForm.js]: added `visibleMetricOptions`

File: src/pages/alerts/components/fields/EvaluatorFields.js [hunk 4]
```diff
@@ -1,116 +1,105 @@
...
   };
-  const getRecommendedValue = () => _.get(RECOMMENDED_METRIC_VALUE, [metric, source, operator], 0);
   const formspec = getFormSpec(metric);
 
   const sourceOptions = formspec.sourceOptions || [];
 
-  const { operatorOptions = [], suffix, sliderLabel, sliderPrecision } = getEvaluatorOptions(metric, source);
+  const { operatorOptions = [], suffix, sliderProps } = getEvaluatorOptions(metric, source);
+
+  const recommendedMetricValue = _.get(RECOMMENDED_METRIC_VALUE, [metric, source, operator]);
```
Comment (bkemper) [inline src/pages/alerts/components/fields/EvaluatorFields.js]: Removed `recommendedMetricValue` default of zero.

File: src/pages/alerts/components/fields/EvaluatorFields.js [hunk 5]
```diff
@@ -1,116 +1,105 @@
...
+      change(FORM_NAME, 'value', RECOMMENDED_METRIC_VALUE[metric][source][event.target.value]);
     }
   };
-  const getRecommendedValue = () => _.get(RECOMMENDED_METRIC_VALUE, [metric, source, operator], 0);
   const formspec = getFormSpec(metric);
 
   const sourceOptions = formspec.sourceOptions || [];
 
-  const { operatorOptions = [], suffix, sliderLabel, sliderPrecision } = getEvaluatorOptions(metric, source);
+  const { operatorOptions = [], suffix, sliderProps } = getEvaluatorOptions(metric, source);
```
Comment (bkemper) [inline src/pages/alerts/components/fields/EvaluatorFields.js]: Rolled up slider props into one variable.

File: src/pages/alerts/components/fields/EvaluatorFields.js [hunk 6]
```diff
@@ -1,116 +1,105 @@
...
+            />
+          </Grid.Column>
+        )}
+        <Grid.Column sm={12} md={sliderLength}>
           <Field
-            name='source'
-            component={SelectWrapper}
+            id="alertEvaluatorValue"
+            name="value"
+            component={SliderFieldWrapper}
```
Comment (bkemper) [inline src/pages/alerts/components/fields/EvaluatorFields.js]: Merged slider and text field into a wrapper.

File: src/pages/alerts/constants/formConstants.js [hunk 7]
```diff
@@ -9,7 +9,8 @@ export const METRICS = {
   health_score: 'Health Score',
   block_bounce_rate: 'Block Bounce Rate',
   hard_bounce_rate: 'Hard Bounce Rate',
-  soft_bounce_rate: 'Soft Bounce Rate'
+  soft_bounce_rate: 'Soft Bounce Rate',
+  injection_count: 'Injection Count',
```
Comment (bkemper) [inline src/pages/alerts/constants/formConstants.js]: added "injection_count"

File: src/pages/alerts/helpers/alertForm.js [hunk 8]
```diff
@@ -42,51 +43,79 @@ const metricToFormSpecMap = {
...
     defaultRecommendedValue: RECOMMENDED_METRIC_VALUE['health_score'].raw.lt,
     defaultFieldValues: [
       { fieldName: 'source', fieldValue: 'raw' },
-      { fieldName: 'operator', fieldValue: 'lt' }
-    ]
-  }
+      { fieldName: 'operator', fieldValue: 'lt' },
+    ],
+  },
+  injection_count: {
```
Comment (bkemper) [inline src/pages/alerts/helpers/alertForm.js]: added "injection_count"

File: src/pages/alerts/helpers/alertForm.js [hunk 9]
```diff
@@ -42,51 +43,79 @@ const metricToFormSpecMap = {
...
+    hasFilters: false,
+    filterType: null,
+    filterOptions: [],
+    sourceOptions: getOptionsFromMap(['raw'], SOURCE_FRIENDLY_NAMES),
+    defaultFieldValues: [
+      { fieldName: 'subaccounts', fieldValue: [] },
+      { fieldName: 'source', fieldValue: 'raw' },
+      { fieldName: 'operator', fieldValue: 'lt' },
+      { fieldName: 'value', fieldValue: 100000 },
```
Comment (bkemper) [inline src/pages/alerts/helpers/alertForm.js]: Set a default, might need to remove.

File: src/pages/alerts/helpers/alertForm.js [hunk 10]
```diff
@@ -42,51 +43,79 @@ const metricToFormSpecMap = {
...
   if (metric in METRICS) {
     return metricToFormSpecMap[metric];
   }
   return {};
 };
 
 export const getEvaluatorOptions = (metric, source) => {
-  const operatorArray = (metric === 'health_score' && source === 'raw') ? ['lt','gt'] : ['gt'];
+  const operatorArray =
+    metric === 'injection_count'
```
Comment (bkemper) [inline src/pages/alerts/helpers/alertForm.js]: added injection_count operator

File: src/pages/alerts/helpers/alertForm.js [hunk 11]
```diff
@@ -42,51 +43,79 @@ const metricToFormSpecMap = {
...
+      ? ['lt']
+      : metric === 'health_score' && source === 'raw'
+      ? ['lt', 'gt']
+      : ['gt'];
   const operatorOptions = getOptionsFromMap(operatorArray, OPERATOR_FRIENDLY_NAMES);
 
-  const suffix = (metric === 'health_score' && source === 'raw') ? '' : '%';
+  const suffix =
+    ['health_score', 'injection_count'].includes(metric) && source === 'raw' ? '' : '%';
```
Comment (bkemper) [inline src/pages/alerts/helpers/alertForm.js]: did not include injection_count suffix

File: src/pages/alerts/helpers/alertForm.js [hunk 12]
```diff
@@ -42,51 +43,79 @@ const metricToFormSpecMap = {
...
+          return { label: 'Score', precision: 1 };
         case 'monthly_sending_limit':
-          return { sliderLabel: 'Percent Used', sliderPrecision: 0 };
+          return { label: 'Percent Used', precision: 0 };
         case 'block_bounce_rate':
         case 'hard_bounce_rate':
         case 'soft_bounce_rate':
-          return { sliderLabel: 'Bounce Percentage Above', sliderPrecision: 0 };
+          return { label: 'Bounce Percentage Above', precision: 0 };
+        case 'injection_count':
```
Comment (bkemper) [inline src/pages/alerts/helpers/alertForm.js]: added injection_count case.

File: src/pages/alerts/components/AlertForm.js [hunk 13]
```diff
@@ -6,25 +6,32 @@ import { TextFieldWrapper, SelectWrapper } from 'src/components';
...
+  { value: '', label: 'Select Metric', disabled: true },
+  ...Object.keys(METRICS).map(key => ({ label: METRICS[key], value: key })),
+];
 
-  resetFormValues = (event) => {
+export class AlertForm extends Component {
+  resetFormValues = event => {
     const { change, isNewAlert, isDuplicate } = this.props;
     const formSpec = getFormSpec(event.target.value);
     const { defaultFieldValues, defaultRecommendedValue } = formSpec;
```
Comment (kchu1337) [inline src/pages/alerts/components/AlertForm.js]: On line 38, you could change to  if `(defaultRecommendedValue  && isNewAlert && !isDuplicate)`.  For injection alerts, the form changes value to undefined (which I guess does nothing?).

File: src/components/reduxFormWrappers/SliderFieldWrapper.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  id,
+  label,
+  max = 100,
+  min = 0,
+  precision = 0,
+  suffix = '',
+  ticks,
+}) => {
+  const cleanAndClamp = React.useCallback(
+    value => {
```
Comment (kchu1337) [inline src/components/reduxFormWrappers/SliderFieldWrapper.js]: value is already used as a prop. Could be confusing.

File: src/pages/alerts/helpers/alertForm.js (hunk 12 above)
Comment (kchu1337) [reply src/pages/alerts/helpers/alertForm.js]: The time it checks this is not clear. But hopefully the pendo tooltip can address that

File: src/pages/alerts/helpers/alertForm.js [hunk 15]
```diff
@@ -42,51 +43,79 @@ const metricToFormSpecMap = {
...
+    return {};
   };
-  const { sliderLabel, sliderPrecision } = getSliderProps();
 
-  return { suffix, operatorOptions, sliderLabel, sliderPrecision };
+  return { suffix, operatorOptions, sliderProps: getSliderProps() };
 };
 
-export const capitalizeFirstLetter = (word) => word.charAt(0).toUpperCase() + word.slice(1);
+export const capitalizeFirstLetter = word => word.charAt(0).toUpperCase() + word.slice(1);
```
Comment (kchu1337) [inline src/pages/alerts/helpers/alertForm.js]: Could replace this usage with `_.upperFirst`, but it's not really in the scope of this ticket.

**Comment by bkemper** [comment]
@kchu1337 This is ready for final review.

**Comment by bkemper** [comment]
@kchu1337 This is ready for final final review.

**Pull request merged**